### PR TITLE
Add lockfile

### PR DESCRIPTION
**Pull request by alexjuda**
- adds `requirements.txt` with dependency pins generated using [pip-tools](https://github.com/jazzband/pip-tools/) by running `$ pip-compile`. This should be used when creating any deployment-related environment (e.g. a docker image) or to help set up a proper environment in development.
- extracts `pytest` to `dev` extra. This way only the project maintainers will install it. Without this change `pytest` and its 
dependencies would pollute the `requirements.txt` file and consequently the target Orquestra docker image

Please regard it as my proposal towards tackling issues highlighted by @simonwa7 and @mstechly in #235.

### CODE REVIEW

File: docker/z-quantum-default/0.0.8/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                mv quilc /usr/local/bin
+RUN git clone https://github.com/rigetti/qvm.git && \
+                cd qvm && \
+                git fetch && \
+                git checkout v1.17.1 && \
+                make QVM_WORKSPACE=10240 qvm && \
+                mv qvm /usr/local/bin
+
+# Install z-quantum-core dependencies
+COPY requirements.txt .
```
[the remark this answers is not in the document]
Comment (alexjuda) [reply docker/z-quantum-default/0.0.8/Dockerfile:60]: Ahh... good catch, I overlooked that. So, do I understand correctly that the original motivation behind this docker image was to preinstall packages that *might* be required by any steps running on Orquestra? 

Also, is there any option for users to select the image that their step runs on?

[the remark this answers is not in the document]
Comment (mstechly) [reply docker/z-quantum-default/0.0.8/Dockerfile:60]: Yes, there is such an option.
Perhaps if we had a pipeline for creating these images automatically, we would have a better system than "one docker image that runs it all", but we don't, so we have this one-size-fits-all solution right now.

Also for more context:
`codecarbon` is used by some code in runtime (though if it's not installed nothing bad happens) and `cvxopt` by code in `z-quantum-qubo` and it is one of the requirements, just takes quite long to install.
Also the part where we install `qvm` and `quilc` is actually used only when you're using `qe-forest`...

[the remark this answers is not in the document]
Comment (mstechly) [reply docker/z-quantum-default/0.0.8/Dockerfile:60]: Actually `cvxpy` not `cvxopt` ;)

[the remark this answers is not in the document]
Comment (alexjuda) [reply docker/z-quantum-default/0.0.8/Dockerfile:66]: I've added the deps you mentioned (f799b22) & updated the comments. Do you think it makes sense now?

[the remark this answers is not in the document]
Comment (mstechly) [reply docker/z-quantum-default/0.0.8/Dockerfile:66]: Looks good, thanks!

**Review by mstechly — APPROVED** [review]
Looks good to me now. 
I think it would be good if @simonwa7 also reviewed it before merging though.

**Review by simonwa7 — APPROVED** [review]

**Comment by simonwa7** [comment]
Looks good to me!

**Pull request merged**